Source/Core/Core/HW/EXI/BBA/BuiltIn.cpp

```diff
@@ -0,0 +1,679 @@
+  const u16 properties = ntohs(tcpdata.properties);
+  if (properties & (TCP_FLAG_FIN | TCP_FLAG_RST))
+  {
+    if (ref == nullptr)
+      return;  // not found
+
+    ref->ack_num++;
+    WriteToQueue(BuildFINFrame(ref));
+    ref->ip = 0;
+    ref->tcp_socket.disconnect();
+  }
+  else if (properties & TCP_FLAG_SIN)
+  {
+    // new connection
+    if (ref != nullptr)
+      return;
+    ref = GetAvailableSlot(0);
+
+    ref->delay = GetTickCountStd();
+    ref->local = tcpdata.source_port;
+    ref->remote = tcpdata.destination_port;
+    ref->ack_num = htonl(tcpdata.sequence_number) + 1;
+    ref->ack_base = ref->ack_num;
+    ref->seq_num = 0x1000000;
+    ref->window_size = htons(tcpdata.window_size);
+    ref->type = IPPROTO_TCP;
+    for (auto& tcp_buf : ref->tcp_buffers)
+      tcp_buf.used = false;
+    ref->from.sin_addr.s_addr = Common::BitCast<u32>(ipdata.destination_addr);
+    ref->from.sin_port = tcpdata.destination_port;
+    ref->to.sin_addr.s_addr = Common::BitCast<u32>(ipdata.source_addr);
+    ref->to.sin_port = tcpdata.source_port;
+    ref->bba_mac = Common::BitCastPtr<Common::MACAddress>(&m_eth_ref->mBbaMem[BBA_NAFR_PAR0]);
+    ref->my_mac = m_fake_mac;
+    ref->tcp_socket.setBlocking(false);
+
+    // reply with a sin_ack
+    Common::TCPPacket result(ref->bba_mac, ref->my_mac, ref->from, ref->to, ref->seq_num,
+                             ref->ack_num, TCP_FLAG_SIN | TCP_FLAG_ACK);
+
+    result.tcp_options = std::vector<u8>{0x02, 0x04, 0x05, 0xb4, 0x01, 0x01, 0x01, 0x01};
+
+    ref->seq_num++;
+    target = sf::IpAddress(htonl(Common::BitCast<u32>(ipdata.destination_addr)));
+    ref->tcp_socket.connect(target, ntohs(tcpdata.destination_port));
+    ref->ready = false;
+    ref->ip = Common::BitCast<u32>(ipdata.destination_addr);
+
+    ref->tcp_buffers[0].data = result.Build();
+    ref->tcp_buffers[0].seq_id = ref->seq_num - 1;
+    ref->tcp_buffers[0].tick = GetTickCountStd() - 900;  // delay
+    ref->tcp_buffers[0].used = true;
+  }
+  else
+  {
+    // data packet
+    if (ref == nullptr)
+      return;  // not found
+
+    const int size = ntohs(ipdata.total_len) - ipdata.DefinedSize() - tcpdata.GetHeaderSize();
+    const u32 this_seq = ntohl(tcpdata.sequence_number);
+
+    if (size > 0)
+    {
+      // only if data
+      if ((int)(this_seq - ref->ack_num) >= 0 && data.size() >= size)
+      {
+        ref->tcp_socket.send(data.data(), size);
+        ref->ack_num += size;
+      }
+
+      // send ack
+      WriteToQueue(BuildAckFrame(ref));
+    }
+    // update windows size
+    ref->window_size = ntohs(tcpdata.window_size);
+
+    // clear any ack data
+    if (tcpdata.properties & ntohs(TCP_FLAG_ACK))
+    {
+      const u32 ack_num = ntohl(tcpdata.acknowledgement_number);
+      for (auto& tcp_buf : ref->tcp_buffers)
+      {
+        if (!tcp_buf.used || tcp_buf.seq_id >= ack_num)
+          continue;
+        Common::TCPHeader* tcppart = (Common::TCPHeader*)&tcp_buf.data[0x22];
```

Comment:
I assume `0x22` (i.e. 34) is `EthernetHeader::SIZE + IPv4Header::SIZE`. `Common::PacketView` should be used to take into account the IPv4 options.